From oliver@neukum.org  Wed Jul 15 06:46:43 2009
From: Oliver Neukum <oliver@neukum.org>
Date: Mon, 13 Jul 2009 10:46:57 +0200
Subject: Staging: serqt_usb2: fix memory leak in error case
To: Greg KH <greg@kroah.com>
Message-ID: <200907131046.57912.oliver@neukum.org>
Content-Disposition: inline


a standard memory leak, as later allocations may fail even if prior
allocations did not. Then the prior allocations must be undone.

Signed-off-by: Oliver Neukum <oliver@neukum.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/staging/serqt_usb2/serqt_usb2.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/drivers/staging/serqt_usb2/serqt_usb2.c
+++ b/drivers/staging/serqt_usb2/serqt_usb2.c
@@ -738,6 +738,11 @@ static int qt_startup(struct usb_serial 
 		if (!qt_port) {
 			dbg("%s: kmalloc for quatech_port (%d) failed!.",
 			    __func__, i);
+			for(--i; i >= 0; i--) {
+				port = serial->port[i];
+				kfree(usb_get_serial_port_data(port));
+				usb_set_serial_port_data(port, NULL);
+			}
 			return -ENOMEM;
 		}
 		spin_lock_init(&qt_port->lock);
